Fix overlay test running on overlay

dmcgowan opened the pull request:
Overlay tests were failing when /var/tmp was an overlay mount with a misleading message.
Now overlay tests will be skipped when attempting to be run on overlay.
Tests will now use the TMPDIR environment variable instead of only /var/tmp

Fixes #21686

thaJeztah commented:
/cc @tonistiigi

anusha-ragunathan commented on daemon/graphdriver/overlay/overlay.go:
```diff
@@ -131,6 +131,9 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
 	case graphdriver.FsMagicZfs:
 		logrus.Error("'overlay' is not supported over zfs.")
 		return nil, graphdriver.ErrIncompatibleFS
+	case graphdriver.FsMagicOverlay:
```
Nit: update comment in L123

anusha-ragunathan commented:
Not this change, but shouldnt aufs's incompatibleFsMagic also include the missing graphdrivers?

tonistiigi commented:
LGTM

dmcgowan committed:
Fix overlay test running on overlay

Overlay tests were failing when /var/tmp was an overlay mount with a misleading message.
Now overlay tests will be skipped when attempting to be run on overlay.
Tests will now use the TMPDIR environment variable instead of only /var/tmp

Fixes #21686

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
Comment updated

vdemeester commented:
LGTM 🐼

karlkfi commented:
So... the layer mount checking method was changed and no longer worked on overlay so you disabled overlay-on-overlay all together rather than make layer mount checking support it?

thaJeztah commented:
@karlkfi is there a specific reason you need to have the `/var/lib/docker` directory in the container's filesystem, and not as a volume? Also note;

> It has never been recommended or supported to run Docker in Docker without running the graph directory in a volume. I have attempted to get around this as well but it has always lead to bad behavior and horrible performance.

karlkfi commented:
@thaJeztah: 

The four places I'm using docker-in-docker (Kubernetes, Mesos, DC/OS, Jenkins) all have the same use case: **Run <controller> in docker and spin up/down new <agents> on-demand, each inside their own docker container with the ability to have their own isolated docker containers.**

This requires the on-demand creation and deletion of self-contained docker environments that can  co-exist with other docker-in-docker environments on the same machine/vm. 

I already have to perform setup pre-allocation steps and pass in information from the host to the parent containers, but I'd like to minimize this because it makes the process more complicated, more fragile, and harder to maintain. The less exceptions I have to code in the entrypoint and pass through via environment variables the better. For example, we already have to pre-allocate IPs for each outer parent container for its child containers to use so that they're accessible to each-other and the host. So the host config specifies IP offsets via environment variables and the outer parent container has a custom entrypoint script that can translate those offsets into docker-engine configuration.

On top of that, I neither need nor want the inner docker environment to persist beyond the lifecycle of the outer container. If I create a volume it persists beyond the lifecycle of the container and has to be separately managed. 

I don't want the inner docker environments to use the outer docker-engine, because:
- the inner and out docker environments should know about each other
- the inner containers should not live beyond the lifecycle of its parent outer container
- the inner container volumes should be host mounted relative to the outer parent container filesystem, not the host filesystem
- the inner containers should not require any cleanup or garbage collection beyond disposing of the outer parent container

karlkfi commented:
For what it's worth, none of the above use cases are used for hosting user-facing production workloads (tho Jenkins lives somewhere in the grey area if you're a platform vendor). So a certain amount of volatility, slowness, and risk is acceptable for the tradeoffs of being orders of magnitude faster and less resource consuming than VMs.  

Docker-in-docker was fine for these use cases up through 1.10, but then this change made it impossible by hard validation without providing a good reason why and without any real alternatives.

thaJeztah commented:
@karlkfi running layered filesystems on top of layered filesystems has always been an issue, and is strongly discouraged, as weird things can happen.

In your case; using an "anonymous" volume should work (i.e. run with `-v /var/lib/docker`), as that would create a new, empty (and randomly named) volume, which can  be deleted together with the container if you `docker rm -v <container>`. Doing so should not require additional preparations. If the container you're running is started from a custom image, you could add

    VOLUME /var/lib/docker

To have docker automatically create a volume

karlkfi commented:
@thaJeztah: I understand it's not supported. But it's never going to get better if it's disabled with no way to even enable it for testing.

As for an anonymous volumes, thanks for the tip, I'll try it out. Is there an equivalent way to delete anonymous volumes when using docker-compose (if not that's kind of a blocker)? Seems like it'd be better to be able to mark a volume for auto-cleanup when creating it tho, rather than needing a special shutdown command.

dmcgowan commented:
@karlkfi I tried to find something to help your out for specifying volumes as temporary for auto cleanup. I do think there might be something missing to enable this. Through scripting this certainly possible by creating volume, running container with "--rm" then deleting volume. It is always possible to mount a temporary directory to the graph directory as well and manage cleanup yourself.

When I have tried to get around this restriction before it was mostly so that I could start up a container which already had images populated, I discovered that was a bad idea for performance and difficult for us to support in terms of managing whiteout files intended to be visible inside the container. In your use case it doesn't sound like having layered content inside the image is what you need, so using a volume will save you from encountering unnecessary performance issues.

karlkfi commented:
Yeah, rewriting docker-compose isn't really what I had in mind...

It sounds like docker needs a way to mark a volume as lifecycle bound to a container. That seems like a better solution than updating docker-compose to know how/when to auto-delete volumes.

karlkfi commented:
Ok, I made a feature request issue to discuss options: https://github.com/docker/docker/issues/27295
